zerver/actions/message_delete.py

```diff
@@ -15,6 +15,34 @@ class DeleteMessagesEvent(TypedDict, total=False):
     stream_id: int
 
 
+def check_update_first_message_id(
+    realm: Realm, stream: Stream, message_ids: List[int], users_to_notify: Iterable[int]
+) -> None:
+    # This will not update the `first_message_id` of streams where the
+    # first message was deleted prior to the implementation of this function.
+    assert stream.recipient_id is not None
+    if stream.first_message_id not in message_ids:
+        return
+    current_first_message_id = (
+        Message.objects.filter(realm_id=realm.id, recipient_id=stream.recipient_id)
+        .values_list("id", flat=True)
+        .order_by("id")
+        .first()
+    )
+
+    stream.first_message_id = current_first_message_id
+    stream.save(update_fields=["first_message_id"])
```

Comment:
This is technically an API change, and deserves a mention in the API changelog (and feature level bump, etc).

Also, can you talk about how you've tested this end-to-end -- does it in fact work as intended?

Reply:
Actually, I was not able to reproduce the UI issue mentioned [here](https://github.com/zulip/zulip/issues/28877#issue-2126203560) but this keeps the `first_message_id` of streams in consistent state. But I am not sure if the extra queries are worth adding if there is no change from a user perspective.
What do you think?

Reply:
So to be clear, are you saying you've checked by print-debugging the web app that `first_message_id` is indeed updated on the web? It's really helpful to be specific about how you tested something, since it saves time for others who might want to try a different form of test.

I can do the final reproduction testing once the API changelog work is ready.

Reply:
> So to be clear, are you saying you've checked by print-debugging the web app that first_message_id is indeed updated on the web?

Yeah, exactly.

Reply:
I have updated the API changelog and `**Changes**` section of the API documentation of corresponding endpoints in `zulip.yaml`.